jidea-69 real time-series values from model run

EmmaLRussell opened the pull request:
Calls the daedalus package to return real time series results. 

The country, pathogen, and response parameters are all passed to the daedalus method. Still to do: hospital capacity and global vaccine investment. 

Currently returning asymptomatic and symptomatic infections, recovered, hospital occupancy, and cumulative deaths. These are all taken directly from the relevant model. compartments. Still to do: vaccinations and compartment delta values e.g. new infections. 

Real run id and parameter values are also being returned in the results. 

All other results values (costs, interventions) are still just dummy values. 

Run requests now validate that expected parameters are provided in the request.

Hardcoded pathogens now fixed to match those in daedalus - we'll validate these values against daedalus-supported values in a future ticket. 

To test: 
1. Run the api:   `./docker/run_containers` 
2.  Request model run: 
```
curl -H "Content-Type: application/json" -d '{"modelVersion": "0.0.1", "parameters": {"country": "Canada", "pathogen": "sars_cov_1", "response": "none", "vaccine": "low", "hospital_capacity": "10500"}}' http://localhost:8001/scenario/run
```
3. Get results:  ` curl http://localhost:8001/scenario/results/[RUN_ID]` passing the run id returned in 2.

EmmaLRussell commented on R/model_run.R:
```diff
@@ -1,4 +1,37 @@
+      values_from = "value",
+      names_from = "compartment"
+    )
+  time_series <- time_series[, c("infect_asymp",
+                                "infect_symp",
+                                "hospitalised",
+                                "recovered",
+                                "dead")]
```
I think those would be useful, yes, if they can be passed through to the front end. One less thing needing to be reproduced in the metadata!

EmmaLRussell commented on R/model_run.R:
```diff
@@ -1,4 +1,37 @@
+  # TODO: include vaccine in params to daedalus
+  vaccine <- parameters$vaccine
+  # TODO: include hospital capacity in params to daedalus
+  hospital_capacity <- parameters$hospital_capacity
```
Up to you! I think the api is the place for the translation from web app terminology to package terminology to take place, so it's probably ok for there to be some differences. But I guess in general it's good to keep them harmonised where possible, especially if we're imagining a bit of a pipeline of web app users becoming package users.

david-mears-2 replied:
I'm confused how these two terms could refer to the same thing - a capacity seems different from a threshold.

david-mears-2 replied:
I see the documentation of the parameter here: https://github.com/jameel-institute/daedalus/pull/27/files#diff-26b50e25422503756a8c78207bc54269300bdedbb5c122249cd8d0e9e1b7c8f3

It seems that response_threshold uses (spare) hospital capacity as a fallback value, so I'd say keeping the name "response threshold" is more accurate, since a government might (probably would) choose to set their threshold differently from the spare hospital capacity.

david-mears-2 commented:
In the last DAEDALUS dashboard meeting we agreed that we don't want to disaggregate asymptomatic and symptomatic infections. We also haven't listed 'recovered' as one of the timeseries that we want to display.

I think the API is probably the right place to configure 'which timeseries to expose in the app', right? So let's exclude 'recovered' here, and sum the (a)symptomatic infections here too.

david-mears-2 commented on inst/json/metadata_0.1.0.json:
```diff
@@ -75,9 +75,11 @@
+      { "id": "dead", "label": "Dead", "description": "Cumulative deaths" }
     ]
   }
 }
```
I've noticed that in this metadata json, we are inconsistent about case - we mix `My Label` and `My label`. My preference is for the latter: capitalise only the first word.

david-mears-2 commented on inst/json/metadata_0.1.0.json:
```diff
@@ -75,9 +75,11 @@
-      { "id": "hosp_occ", "label": "Hospital Occupancy", "description": "Number of hospitalised cases" }
+      { "id": "infect_asymp", "label": "Asymptomatic Infections", "description": "Infections without symptoms" },
+      { "id": "infect_symp", "label": "Symptomatic Infections", "description": "Infections with symptoms" },
+      { "id": "hospitalised", "label": "Hospital Occupancy", "description": "Number of hospitalised cases" },
```
In our results, we get stuff like at a given timestep, several million 'hospitalised', which is far more than the hospital surge capacity (a few tens of thousands). In these cases, calling it hospital 'occupancy', or indeed number of 'hospitalised' cases, isn't quite correct, but I don't have a better suggestion. This naming issue derives from the model itself, of course, which has a hospitalisation compartment that (I guess) isn't capped by the actual hospital capacity - I am assuming 'hospitalisation' basically refers to a severity of illness.

david-mears-2 replied:
@pratikunterwegs Is my assumption correct, that the compartment called 'hospitalised' is more about the number of people with a serious illness, than being about the number of people located in a hospital?

pratikunterwegs replied:
We assume it's serious illness requiring hospitalisation and theoretically, located in a hospital. When the number of hospitalised exceeds the spare capacity, the mortality rate is increased reflecting the pressure on the medical system.

EmmaLRussell commented on inst/json/metadata_0.1.0.json:
```diff
@@ -75,9 +75,11 @@
-      { "id": "cases", "label": "Cases", "description": "Number of confirmed cases" },
-      { "id": "vacc", "label": "Vaccinated", "description": "Cumulative vaccinations" },
-      { "id": "hosp_occ", "label": "Hospital Occupancy", "description": "Number of hospitalised cases" }
+      { "id": "infect_asymp", "label": "Asymptomatic Infections", "description": "Infections without symptoms" },
+      { "id": "infect_symp", "label": "Symptomatic Infections", "description": "Infections with symptoms" },
+      { "id": "hospitalised", "label": "Hospital Occupancy", "description": "Number of hospitalised cases" },
+      { "id": "recovered", "label": "Recovered", "description": "Recovered from infection" },
+      { "id": "dead", "label": "Dead", "description": "Cumulative deaths" }
```
I'll add a ticket for this. Am composing a list of things still todo, and will be creating a bunch of tickets later today, will include this.

EmmaLRussell commented:
I think I've addressed all comments, but see remark on Slack channel - Thailand no longer seems to be on list of package's supported countries. I've set default to UK for now and will add a ticket to reinstate Thailand when available.

david-mears-2 reviewed (approved):
I’ve spent a good while trying to think of a terse but accurate title for the ‘hospitalised’ compartment (which refers to cases that would be hospitalised as long as there is capacity), and the best improvement I’ve come up with is “Hospitalisation demand” rather than “Hospitalised”: has slightly less implication that the cases are in fact successfully in a hospital. Also ran: “Healthcare demand”, “Acute cases”, “Hospital demand”

How do you find that suggestion? As you said, these labels may well get refined further once we share a demo.

david-mears-2 commented on inst/json/metadata_0.1.0.json:
```diff
@@ -60,24 +60,24 @@
     "capacities": [
-      { "id": "hospital_capacity", "label": "Hospital Capacity", "description": "Total hospital beds" },
-      { "id": "icu_capacity", "label": "ICU Capacity", "description": "ICU hospital beds" }
+      { "id": "hospital_capacity", "label": "Hospital capacity", "description": "Total hospital beds" },
```
@pratikunterwegs  Is the hospital capacity that we get as an output similarly about ‘spare’ capacity, as the input parameter is, or is it in fact a total capacity?

david-mears-2 commented on inst/json/metadata_0.1.0.json:
```diff
@@ -60,24 +60,24 @@
-      { "id": "cases", "label": "Cases", "description": "Number of confirmed cases" },
-      { "id": "vacc", "label": "Vaccinated", "description": "Cumulative vaccinations" },
-      { "id": "hosp_occ", "label": "Hospital Occupancy", "description": "Number of hospitalised cases" }
+      { "id": "infect", "label": "Infections", "description": "Infections not requiring hospitalisation" },
```
Wondering about the possibility of a more accurate title for Infections too — best I have is “community infections”, what do you think?
